### PR TITLE
new CustomModelData expressions/condition

### PR DESCRIPTION
**Pull request by ShaneBeee**
### Description
This adds 2 new expressions and 1 condition for 1.14's CustomModelData tag on items.
**What is CustomModelData?**
It's a new tag added to items in 1.14. It allows for items to have different textures set for them, allowing each item to operate independently while still retaining their durability. 
On previous version of the game (1.13.2 and below) custom models were done by creating an **unbreakable** item, and setting their durability. Whilst this method does work, the issue came down to the item being unbreakable, therefor not showing a durability indicator on said item. Also to note this could only be used on items that could be damaged (ie: tools/weapons, not food).
1.14's CustomModelData tag can be used on ANY item, regardless if it holds durability or not.

**Expressions**:
- item with custom model data = This allows to give an item to a player with custom model data attached to it, ex:
`give player a wooden hoe with custom model data 1`
`give player a diamond sword named "Emerald Sword" with custom model data 1`
- custom model data of item = This allows to get/add/set/remove/reset the custom model data of an item, ex:
`add 1 to custom model data of player's tool`
`set custom model data of player's tool to 10`
`set {_data} to custom model data of player's tool`

**Condition**:
- has custom model data = This allows you to check if an item currently has the custom model data tag set, ex:
`if player's tool has custom model data:`
---
**Target Minecraft Versions:** 1.14+
**Requirements:** Spigot 1.14+
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/conditions/CondHasCustomModelData.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+@Name("Has Custom Model Data")
+@Description("Check if an item has a custom model data tag")
+@Examples("player's tool has custom model data")
+@Since("INSERT VERSION")
+public class CondHasCustomModelData extends PropertyCondition<ItemStack> {
+	
+	static {
+		if (Skript.methodExists(ItemMeta.class, "hasCustomModelData")) {
+			register(CondHasCustomModelData.class, PropertyType.HAVE, "[custom] model data", "itemstacks");
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/conditions/CondHasCustomModelData.java]: Should be taking ItemType here

File: src/main/java/ch/njol/skript/expressions/ExprCustomModelData.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+@Description("Get/set the CustomModelData tag for an item. (Value is an integer between 0 and 99999999)")
+@Examples({"set custom model data of player's tool to 3",
+	"set {_model} to custom model data of player's tool"})
+@RequiredPlugins("1.14+")
+@Since("INSERT VERSION")
+public class ExprCustomModelData extends SimplePropertyExpression<ItemStack, Long> {
+	
+	static {
+		if (Skript.methodExists(ItemMeta.class, "hasCustomModelData")) {
+			register(ExprCustomModelData.class, Long.class, "[custom] model data", "itemstack");
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprCustomModelData.java]: Same thing here, also, why does it not take multiple items?

File: src/main/java/ch/njol/skript/conditions/CondHasCustomModelData.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import ch.njol.skript.conditions.base.PropertyCondition;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+
+@Name("Has Custom Model Data")
+@Description("Check if an item has a custom model data tag")
+@Examples("player's tool has custom model data")
+@Since("INSERT VERSION")
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/conditions/CondHasCustomModelData.java]: Forgot the RequiredPlugins annotation here

File: src/main/java/ch/njol/skript/conditions/CondHasCustomModelData.java (hunk 1 above)
Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/conditions/CondHasCustomModelData.java]: I had a really really hard time creating this with item type, when I first made it a few months ago in Sk-NBeeT, but I will attempt again.

File: src/main/java/ch/njol/skript/conditions/CondHasCustomModelData.java (hunk 3 above)
Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/conditions/CondHasCustomModelData.java]: Yeah, I noticed that after, I already have the change I just need to push it. Will do shortly when I fix the other things. :)

File: src/main/java/ch/njol/skript/conditions/CondHasCustomModelData.java (hunk 1 above)
Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/conditions/CondHasCustomModelData.java]: fixed

File: src/main/java/ch/njol/skript/expressions/ExprCustomModelData.java (hunk 2 above)
Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/expressions/ExprCustomModelData.java]: fixed, and now takes multiple items

File: src/main/java/ch/njol/skript/expressions/ExprCustomModelData.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+			register(ExprCustomModelData.class, Long.class, "[custom] model data", "itemtypes");
+		}
+	}
+	
+	@SuppressWarnings("null")
+	@Override
+	public Long convert(ItemType item) {
+		assert item.getItemMeta() != null;
+		if (item.getItemMeta().hasCustomModelData())
+			return (long) item.getItemMeta().getCustomModelData();
```
Comment (Wealthyturtle) [inline src/main/java/ch/njol/skript/expressions/ExprCustomModelData.java]: Minor nitpick, but it pains me to see 3 different calls to getItemMeta.

Use
```
		ItemMeta meta = item.getItemMeta();
		assert meta != null;
		if (meta.hasCustomModelData())
			return (long) meta.getCustomModelData();
```
Instead

Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/expressions/ExprCustomModelData.java]: Sometimes I wonder about myself.... I'll fix that right up!

**Comment by bensku** [comment]
We can't test interaction with MC client automatically, but testing that CustomModelData is set/get correctly would be good.

**Comment by ShaneBeee** [comment]
> We can't test interaction with MC client automatically, but testing that CustomModelData is set/get correctly would be good.

Ok, I added a test, but I wasn't 100% sure if this one was supposed to go in "regression" or not. Let me know if that needs to be changed.

**Comment by bensku** [comment]
This does not fix a bug, so it is not a regression test. Refer to test system README for more details.

**Comment by ShaneBeee** [comment]
> This does not fix a bug, so it is not a regression test. Refer to test system README for more details.

Ah yes, sorry about that. I keep forgetting the tests has its own readme. Ill switch that right up

**Pull request merged**